File: measuretool/MeasureTools/CalculateRangeAtY.m
function [x1,x2] = CalculateRangeAtY(elpparm,y)
% 计算椭圆一般方程当y值已知时，计算出的两个交点，如果不存在则为空集
elpparm = double(elpparm);
A = elpparm(1); B = elpparm(2); C = elpparm(3); D = elpparm(4); E = elpparm(5); F = elpparm(6);
Delta = (B*y+D)^2 - A*(C*y^2+2*E*y+F);
if Delta< 0 
    x1 = []; x2 = [];
else
    x1 = (-(B*y+D)-sqrt(Delta))/A;
    x2 = (-(B*y+D)+sqrt(Delta))/A;
%     x1 = round(x1); x2 = round(x2);
    
    if x2 < x1
        t = x1;
        x1 = x2;
        x2 = t;
    end
    
end



end

File: measuretool/MeasureTools/CalculateRangeAtY.cpp
#include "MeasureTools.h"

#include <iostream>

using namespace std;

void CalculateRangeAtY(double *elpparm, double y, double *x1, double *x2)
{
	double A, B, C, D, E, F;
	A = elpparm[0], B = elpparm[1], C = elpparm[2];
	D = elpparm[3], E = elpparm[4], F = elpparm[5];

	double Delta = pow(B*y + D, 2) - A*(C*y*y + 2 * E*y + F);

	if (Delta < 0)
		*x1 = -10, *x2 = -20;
	else
	{
        double t1, t2;
		t1 = (-(B*y + D) - sqrt(Delta)) / A;
		t2 = (-(B*y + D) + sqrt(Delta)) / A;
		
		if (t2 < t1)
			swap(t1, t2);
        
        *x1 = t1;
        *x2 = t2;
	}



}